aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -616,9 +619,9 @@ std::tuple<Tensor, Tensor, Tensor> _lu_with_info_cpu(const Tensor& self, bool pi
   }
   if (check_errors) {
     if (self.dim() > 2) {
-      batchCheckErrors(infos_tensor, "lu");
+      batchCheckErrors(infos_tensor, "lu", /*allow_singular=*/true);
```

Comment:
Earlier, if inputs where singular, then we would throw an error about the singularity. What this change does is that it allows singular inputs, but _why_?

Reply:
Doing LU factorization on singular matrices is a valid operation. So, in `lu` we can allow singular inputs, especially, when dealing with rectangular matrixes.
However, in `lu_solve`, singularity is not allowed and error is raised.

Reply:
So, this is a bc-breaking change. I think introducing this behavior is harmful for the following reason:
- MAGMA has a bug wherein if you perform batched LU, the returned LU matrix has `nan`s in it if the input is singular. This might lead to people misconstruing that there's a bug in their code.

cc: @SsnL on bc-breaking change

Reply:
What version of magma are you using?
Magma 1.6.1 release notes contains:
```
Fixing "nan" output for batched factorizations.
```
Can you confirm that this bug has not been fixed?

Reply:
Yeah, this hasn't been fixed:
```python
>>> import torch
>>> a = torch.ones(1, 3, 3, device='cuda')
>>> a.lu(get_infos=True)
(tensor([[[1., 1., 1.],
         [1., 0., 0.],
         [1., nan, nan]]], device='cuda:0'), tensor([[1, 2, 3]], device='cuda:0', dtype=torch.int32), tensor([2], device='cuda:0', dtype=torch.int32))
```

Reply:
However:
```
>>> a.lu(pivot=False)
(tensor([[[1., 1., 1.],
         [1., 0., 0.],
         [1., 0., 0.]]], device='cuda:0'), tensor([[1, 2, 3]], device='cuda:0', dtype=torch.int32))
```
So, the issue seems to be in `magmaLuBatched`.

Reply:
In addition, the lu nan issue exists only for small square matrices (m<=32).
For instance, try `torch.ones(1, 33, 33, device='cuda').lu()` vs `torch.ones(1, 32, 32, device='cuda').lu()`.

Reply:
@vishwakftw I see you have reported related issues to magma forum. I sent a report about this issue to magma forum as well.
However, it looks like magma devs are not very responsive in fixing these issues. Here are some possible approaches to fix this issue (other than just disabling the required feature):
1. fix magma at magma conda package level (that is, add a patch to magma-101, .. recipes). And send the patch to upstream. This might be the fastest approach to achieve the required functionality in full.
2. if m==n and m<=32 then raise an error on a singular input (current behavior in master), otherwise, allow singular inputs (current behavior in this PR). And document this. This is the fastest approach to get the required functionality but only partially.

What do you think?

Reply:
The MAGMA bug that appears only when computing LU factorization of a patch of square singular matrices with sizes less than 33, is now documented in lu docs. This should avoid users misinterpreting the nan values (provided that they are able to track the issue to the lu call).

Reply:
I just uncovered another issue about passing singular matrices:
```python
>>> a = torch.ones(1, 33, 33).cuda()
>>> a.lu(get_infos=True)[1]
tensor([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 33, 34,
         35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]],
       device='cuda:0', dtype=torch.int32)
>>> a.cpu().lu(get_infos=True)[1]
tensor([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
         19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33]],
       dtype=torch.int32)
```

Reply:
good find

Reply:
Resolving as the magma bug (magma issue 13) has been worked around.